src/hotspot/share/opto/mempointer.cpp

```diff
@@ -186,9 +215,69 @@ void MemPointerDecomposedFormParser::parse_sub_expression(const MemPointerSumman
   _summands.push(summand);
 }
 
+bool MemPointerParser::sub_expression_has_native_base_candidate(Node* start) {
+  // BFS over the expression.
+  ResourceMark rm;
+  GrowableArray<Node*> worklist;
+  worklist.append(start);
+  for (int i = 0; i < worklist.length(); i++) {
+    Node* n = worklist.at(i);
+    switch(n->Opcode()) {
+      case Op_AddL:
+        // Traverse to both inputs.
+        worklist.append(n->in(1));
+        worklist.append(n->in(2));
+        break;
+      case Op_SubL:
+      case Op_CastLL:
+        // Traverse to the first input. The base cannot be on the rhs of a sub.
+        worklist.append(n->in(1));
+        break;
+      default:
+        if (is_native_memory_base_candidate(n)) { return true; }
+        break;
+    }
+    // This is a heuristic, so we are allowed to bail out early if the graph
+    // is too deep. The constant is chosen arbitrarily, not too large but big
+    // enough for all normal cases.
+    if (worklist.length() > 100) { return false; }
+  }
+  // Parsed over the whole expression, nothing found.
+  return false;
+}
+
+// Check if the node is a candidate to be a memory segment "base".
+// (1) CastX2P: some arbitrary long that is cast to a pointer.
+// (2) LoadL from field jdk.internal.foreign.NativeMemorySegmentImpl.min
+//     Holds the address() of a native memory segment.
+bool MemPointerParser::is_native_memory_base_candidate(Node* n) {
```

Comment:
Does this really belong to the move from VPointer to MemPointer? AFAIU, it's an extra optimization on top of the move. Should be done as a separate PR?

Reply:
I had thought about splitting this out, but it I don't see a great way of doing that. Let me explain.

Before this patch, `MemPointer` parses through `CastX2P` in all cases, but `VPointer` never parses through `CastX2P`.

The current patch strikes a compromise:
Only parse through `CastX2P` if we have a `MemorySegment` access.

Let me explore some alternatives here:
- Do a separate patch, and make the change in `MemPointer` first, restricting parsing through `CastX2P` to MemorySegment cases. This could be done, but would not make sense on its own yet, as `MemPointer` is perfectly happy with parsing through `CastX2P` in all cases.
- First replace `VPointer` parsing with `MemPointer` logic. But then the question is what I do with the parsing through `CastX2P`.
  - Never parse through CastX2P: I would have to temporarily introduce some flag that is enabled for MergeStores, but disabled for VPointer. That is extra code you would have to review as well.
  - Always parse through CastX2P: that would make it impossible to find a base in non-MemorySegment cases of native memory access. There are some examples in 
  -  `test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMismatchedAccess.java`. I would parse through `CastX2P`, and then it is not clear what would be a suitable "base", which I now need for `VPointer` (MergeStores did not need to know about bases).

TLDR: I'm not sure how to split this out without creating regressions, or creating unmotivated/untested code in the meantime. I also think that this code is relatively contained: it's all in `sub_expression_has_native_base_candidate` and `is_native_memory_base_candidate`.

I was also thinking of doing some refactoring of `mempointer.cpp/hpp` separately first - but again that would not be very well motivated: for example if I add the "Base" logic first separately, then it has no use. It will only be used by `VPointer`. And these changes are very well contained to those files, so if you just look at these files first, and only then at the other files it should be relatively straight forward.

Let me know if you have a good idea though.

Reply:
Thanks for the explanation. I think it's reasonable to keep it in this change then.